File: agent/tools/dataset_write.py
#
#  Copyright 2026 The InfiniFlow Authors. All Rights Reserved.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You can obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#
from __future__ import annotations

import logging
import os
import re
from abc import ABC
from typing import Any

from agent.tools.base import ToolBase, ToolMeta, ToolParamBase
from common.connection_utils import timeout
from common.constants import TaskStatus
from common.data_source.seafile_client import SeafileClient
from common.data_source.seafile_week import (
    WEEK_ID_RE,
    normalize_weekly_output_format,
    render_weekly_report,
    resolve_library_id,
    seafile_publish_configured,
    weekly_report_content_type,
    weekly_report_filename,
)

logger = logging.getLogger(__name__)

FILENAME_RE = re.compile(r"^weekly-report-\d{4}-W\d{2}\.(md|docx|pdf)$")


class _MemoryUpload:
    def __init__(self, filename: str, blob: bytes, doc_id: str | None = None):
        self.filename = filename
        self._blob = blob
        if doc_id:
            self.id = doc_id

    def read(self) -> bytes:
        return self._blob


class DatasetWriteParam(ToolParamBase):
    def __init__(self):
        self.meta: ToolMeta = {
            "name": "write_my_dataset",
            "description": "Write the generated weekly report markdown into the configured dataset. Filename is computed from week_id.",
            "parameters": {
                "content": {
                    "type": "string",
                    "description": "Weekly report markdown.",
                    "default": "",
                    "required": False,
                },
                "week_id": {
                    "type": "string",
                    "description": "ISO week id such as 2026-W33.",
                    "default": "",
                    "required": False,
                },
            },
        }
        super().__init__()
        self.dataset_ids: list[str] = []
        self.kb_ids: list[str] = []
        self.publish_policy = "auto"
        self.language = "zh"
        self.seafile_connector_id = ""
        self.seafile_repo_id = ""
        self.seafile_path = ""
        self.output_format = "md"

    def check(self):
        policy = (self.publish_policy or "auto").lower()
        if policy not in {"auto", "draft"}:
            raise ValueError("publish_policy must be auto or draft")
        self.output_format = normalize_weekly_output_format(getattr(self, "output_format", "md"))

    def get_input_form(self) -> dict[str, dict]:
        return {
            "content": {"name": "Content", "type": "line", "optional": True},
            "week_id": {"name": "Week id", "type": "line", "optional": True},
        }


class DatasetWrite(ToolBase, ABC):
    component_name = "DatasetWrite"

    @property
    def _dataset_ids(self) -> list[str]:
        return self._param.dataset_ids or getattr(self._param, "kb_ids", None) or []

    def _resolve_dataset_id(self) -> str:
        from api.db.services.knowledgebase_service import KnowledgebaseService

        ids = [i for i in self._dataset_ids if i]
        if len(ids) != 1:
            raise ValueError("DatasetWrite requires exactly one dataset_id.")
        dataset_id = ids[0]
        if "@" in dataset_id:
            dataset_id = str(self._canvas.get_variable_value(dataset_id) or dataset_id)
        tenant_id = self._canvas.get_tenant_id()
        if not KnowledgebaseService.accessible(dataset_id, tenant_id):
            raise PermissionError("Weekly-report dataset is not accessible to this canvas owner.")
        return dataset_id

    def _resolve_text(self, value: Any) -> str:
        text = "" if value is None else str(value).strip()
        if text.startswith("{") or "@" in text:
            text = str(self._canvas.get_variable_value(text) or "").strip()
        return text

    def _write_seafile(self, filename: str, content: bytes) -> dict[str, Any] | None:
        repo = self._resolve_text(getattr(self._param, "seafile_repo_id", ""))
        path = self._resolve_text(getattr(self._param, "seafile_path", ""))
        if not seafile_publish_configured(repo, path):
            return None
        connector_id = self._resolve_text(getattr(self._param, "seafile_connector_id", ""))
        from api.db.services.connector_service import ConnectorService

        tenant_id = self._canvas.get_tenant_id()
        connector = None
        if connector_id:
            if not ConnectorService.accessible(connector_id, tenant_id):
                raise PermissionError("Seafile connector is not accessible to this canvas owner.")
            ok, connector = ConnectorService.get_by_id(connector_id)
            if not ok:
                raise ValueError(f"Seafile connector {connector_id} was not found.")
        else:
            for row in ConnectorService.query(tenant_id=tenant_id, source="seafile") or []:
                connector = row
                break
        if connector is None:
            raise ValueError("已填写 Seafile 资料库和路径，但没有可用的 Seafile 数据源。")
        source = (getattr(connector, "source", "") or "").lower()
        if source != "seafile":
            raise ValueError(f"Connector {getattr(connector, 'id', '')} is source={source!r}, expected seafile.")
        config = connector.config or {}
        credentials = config.get("credentials") or {}
        client = SeafileClient(
            config.get("seafile_url") or "",
            token=credentials.get("seafile_token") or config.get("seafile_token"),
            repo_token=credentials.get("repo_token") or config.get("repo_token"),
            download_hosts=list(config.get("download_hosts") or []),
        )
        repo_id = resolve_library_id(client.list_libraries(), repo)
        uploaded = client.upsert_file(
            repo_id,
            path,
            filename,
            content,
            content_type=weekly_report_content_type(self._output_format()),
        )
        uploaded["repo_id"] = repo_id
        return uploaded

    def _output_format(self) -> str:
        return normalize_weekly_output_format(getattr(self._param, "output_format", "md"))

    def _report_file(self, week_id: str, content: Any) -> tuple[str, bytes]:
        if content is None:
            content = ""
        if not isinstance(content, str):
            content = str(content)
        fmt = self._output_format()
        filename = weekly_report_filename(week_id, fmt)
        if not FILENAME_RE.fullmatch(filename):
            raise ValueError(f"computed filename rejected: {filename}")
        return filename, render_weekly_report(content, fmt)

    def _resolve_week_id(self, kwargs: dict[str, Any]) -> str:
        week_id = (kwargs.get("week_id") or getattr(self._param, "week_id", "") or "").strip()
        if "@" in week_id or week_id.startswith("{"):
            week_id = str(self._canvas.get_variable_value(week_id) or "")
        if not WEEK_ID_RE.fullmatch(week_id):
            raise ValueError(f"DatasetWrite week_id must look like YYYY-Www, got {week_id!r}")
        return week_id

    @timeout(int(os.environ.get("COMPONENT_EXEC_TIMEOUT", 120)))
    def _invoke(self, **kwargs):
        if self.check_if_canceled("DatasetWrite processing"):
            return

        policy = (self._param.publish_policy or "auto").lower()
        if policy == "draft":
            result = {"action": "skipped", "reason": "publish_policy=draft"}
            self.set_output("json", result)
            self.set_output("formalized_content", "DatasetWrite skipped (draft).")
            return result

        try:
            from api.db.services.document_service import DocumentService
            from api.db.services.doc_metadata_service import DocMetadataService
            from api.db.services.file_service import FileService
            from api.db.services.knowledgebase_service import KnowledgebaseService

            dataset_id = self._resolve_dataset_id()
            week_id = self._resolve_week_id(kwargs)
            filename, blob = self._report_file(week_id, kwargs.get("content"))
            fmt = self._output_format()

            ok, kb = KnowledgebaseService.get_by_id(dataset_id)
            if not ok:
                raise ValueError(f"Dataset {dataset_id} was not found.")

            existing = list(DocumentService.query(name=filename, kb_id=dataset_id))
            if len(existing) > 1:
                raise RuntimeError(f"ambiguous_name: {len(existing)} documents named {filename} in dataset {dataset_id}")

            tenant_id = self._canvas.get_tenant_id()
            if existing:
                upload = _MemoryUpload(filename, blob, doc_id=existing[0].id)
            else:
                upload = _MemoryUpload(filename, blob)

            err, files = FileService.upload_document(kb, [upload], tenant_id)
            if err:
                raise RuntimeError("; ".join(err))
            if not files:
                raise RuntimeError("DatasetWrite upload returned no document.")

            doc, _stored = files[0]
            doc_id = doc["id"] if isinstance(doc, dict) else doc.id
            DocMetadataService.update_document_metadata(
                doc_id,
                {
                    "week_id": week_id,
                    "language": self._param.language or "zh",
                    "kind": "weekly_report",
                    "output_format": fmt,
                },
            )
            ok, fresh = DocumentService.get_by_id(doc_id)
            if ok:
                DocumentService.update_by_id(
                    doc_id,
                    {
                        "run": TaskStatus.UNSTART.value,
                        "progress": 0,
                        "progress_msg": "",
                    },
                )
                ok, fresh = DocumentService.get_by_id(doc_id)
                DocumentService.run(tenant_id, fresh.to_dict(), {})

            result = {
                "action": "replaced" if existing else "created",
                "document_id": doc_id,
                "filename": filename,
                "week_id": week_id,
                "dataset_id": dataset_id,
                "output_format": fmt,
                "ingest": "queued",
            }
            seafile = self._write_seafile(filename, blob)
            if seafile:
                result["seafile"] = seafile
                summary = f"Wrote {filename} ({result['action']}). Seafile {seafile.get('path')} ({seafile.get('action')})."
            else:
                result["seafile"] = {"action": "skipped"}
                summary = f"Wrote {filename} ({result['action']})."
            self.set_output("json", result)
            self.set_output("formalized_content", summary)
            return result
        except Exception as exc:
            logger.exception("DatasetWrite failed")
            self.set_output("_ERROR", str(exc))
            self.set_output("json", {"action": "error", "error": str(exc)})
            self.set_output("formalized_content", "")
            return str(exc)

File: internal/agent/tool/dataset_write.go
package tool

import (
	"archive/zip"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"html"
	"regexp"
	"strings"
	"sync"

	"github.com/cloudwego/eino/components/tool"
	"github.com/cloudwego/eino/schema"

	"ragflow/internal/agent/runtime"
	seafilepkg "ragflow/internal/connector/seafile"
)

const datasetWriteToolName = "write_my_dataset"

var weeklyFilenameRE = regexp.MustCompile(`^weekly-report-\d{4}-W\d{2}\.(md|docx|pdf)$`)
var weekIDRE = regexp.MustCompile(`^\d{4}-W\d{2}$`)

type DatasetWriteRequest struct {
	DatasetID string
	WeekID    string
	Filename  string
	Content   string
	Body      []byte
	Metadata  map[string]any
}

type DatasetWriteResult struct {
	Action     string
	DocumentID string
	Filename   string
}

// DatasetWriter upserts a markdown weekly report by exact name.
type DatasetWriter func(ctx context.Context, req DatasetWriteRequest) (*DatasetWriteResult, error)

var (
	datasetWriterMu sync.RWMutex
	datasetWriter   DatasetWriter
)

func SetDatasetWriter(fn DatasetWriter) {
	datasetWriterMu.Lock()
	defer datasetWriterMu.Unlock()
	datasetWriter = fn
}

type datasetWriteParams struct {
	DatasetIDs         []string
	PublishPolicy      string
	Language           string
	OutputFormat       string
	SeafileConnectorID string
	SeafileRepoID      string
	SeafilePath        string
}

type DatasetWriteTool struct {
	defaults datasetWriteParams
}

func NewDatasetWriteTool() *DatasetWriteTool {
	return NewDatasetWriteToolWithDefaults(datasetWriteParams{PublishPolicy: "auto", Language: "zh"})
}

func NewDatasetWriteToolWithDefaults(defaults datasetWriteParams) *DatasetWriteTool {
	if defaults.PublishPolicy == "" {
		defaults.PublishPolicy = "auto"
	}
	if format, err := normalizeWeeklyOutputFormat(defaults.OutputFormat); err == nil {
		defaults.OutputFormat = format
	} else {
		defaults.OutputFormat = "md"
	}
	return &DatasetWriteTool{defaults: defaults}
}

func (d *DatasetWriteTool) Info(_ context.Context) (*schema.ToolInfo, error) {
	return &schema.ToolInfo{
		Name: datasetWriteToolName,
		Desc: "Write the generated weekly report markdown into the configured dataset.",
		ParamsOneOf: schema.NewParamsOneOfByParams(map[string]*schema.ParameterInfo{
			"content": {Type: schema.String, Desc: "Weekly report markdown", Required: false},
			"week_id": {Type: schema.String, Desc: "ISO week id such as 2026-W33", Required: false},
		}),
	}, nil
}

func (d *DatasetWriteTool) InvokableRun(ctx context.Context, argsJSON string, _ ...tool.Option) (string, error) {
	args := map[string]any{}
	if strings.TrimSpace(argsJSON) != "" {
		if err := json.Unmarshal([]byte(argsJSON), &args); err != nil {
			return datasetWriteErrJSON(err), nil
		}
	}
	if strings.EqualFold(d.defaults.PublishPolicy, "draft") {
		raw, _ := json.Marshal(map[string]any{"json": map[string]any{"action": "skipped", "reason": "publish_policy=draft"}})
		return string(raw), nil
	}
	weekID, _ := args["week_id"].(string)
	weekID = strings.TrimSpace(weekID)
	if !weekIDRE.MatchString(weekID) {
		return datasetWriteErrJSON(fmt.Errorf("week_id must look like YYYY-Www")), nil
	}
	content, _ := args["content"].(string)
	filename, blob, err := buildWeeklyReportFile(weekID, content, d.defaults.OutputFormat)
	if err != nil {
		return datasetWriteErrJSON(err), nil
	}
	if !weeklyFilenameRE.MatchString(filename) {
		return datasetWriteErrJSON(fmt.Errorf("computed filename rejected")), nil
	}
	if len(d.defaults.DatasetIDs) != 1 {
		return datasetWriteErrJSON(fmt.Errorf("DatasetWrite requires exactly one dataset_id")), nil
	}
	format, _ := normalizeWeeklyOutputFormat(d.defaults.OutputFormat)
	datasetWriterMu.RLock()
	writer := datasetWriter
	datasetWriterMu.RUnlock()
	if writer == nil {
		return datasetWriteErrJSON(fmt.Errorf("dataset writer is not wired; use the Python canvas for scheduled publish")), nil
	}
	result, err := writer(ctx, DatasetWriteRequest{
		DatasetID: d.defaults.DatasetIDs[0],
		WeekID:    weekID,
		Filename:  filename,
		Content:   content,
		Body:      blob,
		Metadata:  map[string]any{"week_id": weekID, "kind": "weekly_report", "language": d.defaults.Language, "output_format": format},
	})
	if err != nil {
		return datasetWriteErrJSON(err), nil
	}
	payload := map[string]any{
		"action":        result.Action,
		"document_id":   result.DocumentID,
		"filename":      result.Filename,
		"week_id":       weekID,
		"output_format": format,
	}
	summary := fmt.Sprintf("Wrote %s (%s).", result.Filename, result.Action)
	if seafileConfigured(d.defaults.SeafileRepoID, d.defaults.SeafilePath) {
		seafileResult, seafileErr := publishWeeklyReportToSeafile(ctx, d.defaults, filename, blob)
		if seafileErr != nil {
			return datasetWriteErrJSON(seafileErr), nil
		}
		payload["seafile"] = seafileResult
		summary = fmt.Sprintf("%s Seafile %v (%v).", summary, seafileResult["path"], seafileResult["action"])
	} else {
		payload["seafile"] = map[string]any{"action": "skipped"}
	}
	raw, _ := json.Marshal(map[string]any{
		"json":               payload,
		"formalized_content": summary,
	})
	return string(raw), nil
}

func seafileConfigured(repoID, path string) bool {
	return strings.TrimSpace(repoID) != "" && strings.TrimSpace(path) != ""
}

func publishWeeklyReportToSeafile(ctx context.Context, defaults datasetWriteParams, filename string, content []byte) (map[string]any, error) {
	connectorID := strings.TrimSpace(defaults.SeafileConnectorID)
	if connectorID == "" {
		return nil, fmt.Errorf("seafile connector_id is required when library and path are set")
	}
	userID := ""
	if state, _, err := runtime.GetStateFromContext[*runtime.CanvasState](ctx); err == nil && state != nil {
		userID = strings.TrimSpace(fmt.Sprint(state.Sys["user_id"]))
		if userID == "" {
			userID = strings.TrimSpace(fmt.Sprint(state.Sys["tenant_id"]))
		}
	}
	snapshot, err := lookupConnector(ctx, connectorID, userID)
	if err != nil {
		return nil, err
	}
	if snapshot == nil || !strings.EqualFold(snapshot.Source, "seafile") {
		return nil, fmt.Errorf("connector is not a seafile source")
	}
	cfg := snapshot.Config
	if cfg == nil {
		cfg = map[string]any{}
	}
	creds, _ := cfg["credentials"].(map[string]any)
	if creds == nil {
		creds = map[string]any{}
	}
	token, _ := creds["seafile_token"].(string)
	repoToken, _ := creds["repo_token"].(string)
	baseURL, _ := cfg["seafile_url"].(string)
	client := seafilepkg.NewClient(baseURL, token, repoToken, nil)
	libraries, err := client.ListLibraries()
	if err != nil {
		return nil, err
	}
	repoID, err := resolveSeafileLibraryID(libraries, defaults.SeafileRepoID)
	if err != nil {
		return nil, err
	}
	uploaded, err := client.UpsertFile(repoID, defaults.SeafilePath, filename, content)
	if err != nil {
		return nil, err
	}
	uploaded["repo_id"] = repoID
	return uploaded, nil
}

func resolveSeafileLibraryID(libraries []map[string]any, repo string) (string, error) {
	wanted := strings.TrimSpace(repo)
	if wanted == "" {
		return "", fmt.Errorf("seafile library is empty")
	}
	for _, lib := range libraries {
		id, _ := lib["id"].(string)
		if strings.TrimSpace(id) == wanted {
			return wanted, nil
		}
	}
	for _, lib := range libraries {
		name, _ := lib["name"].(string)
		id, _ := lib["id"].(string)
		if strings.TrimSpace(name) == wanted && id != "" {
			return id, nil
		}
	}
	return "", fmt.Errorf("seafile library %q was not found", wanted)
}

func datasetWriteErrJSON(err error) string {
	raw, _ := json.Marshal(map[string]any{"_ERROR": err.Error(), "json": map[string]any{"action": "error"}})
	return string(raw)
}

func (d *DatasetWriteTool) ComponentSpec() ComponentSpec {
	return ComponentSpec{
		Inputs: map[string]string{
			"content": "Weekly report markdown",
			"week_id": "ISO week id",
		},
		Outputs: map[string]string{
			"json":               "Publish result",
			"formalized_content": "Human-readable publish summary",
		},
		InputForm: map[string]any{
			"content": map[string]any{"type": "line", "name": "Content"},
			"week_id": map[string]any{"type": "line", "name": "Week id"},
		},
	}
}

func (d *DatasetWriteTool) BuildComponentOutputs(envelope map[string]any) map[string]any {
	out := map[string]any{}
	for _, key := range []string{"json", "formalized_content", "_ERROR"} {
		if v, ok := envelope[key]; ok {
			out[key] = v
		}
	}
	return out
}

func normalizeWeeklyOutputFormat(value string) (string, error) {
	key := strings.TrimSpace(strings.ToLower(value))
	if key == "" {
		return "md", nil
	}
	key = strings.TrimPrefix(key, ".")
	switch key {
	case "md", "markdown":
		return "md", nil
	case "docx", "doc":
		return "docx", nil
	case "pdf":
		return "pdf", nil
	default:
		return "", fmt.Errorf("output_format must be one of md, docx, pdf, got %q", value)
	}
}

func weeklyReportFilename(weekID, outputFormat string) (string, error) {
	if !weekIDRE.MatchString(weekID) {
		return "", fmt.Errorf("week_id must look like YYYY-Www, got %q", weekID)
	}
	format, err := normalizeWeeklyOutputFormat(outputFormat)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("weekly-report-%s.%s", weekID, format), nil
}

func buildWeeklyReportFile(weekID, content, outputFormat string) (string, []byte, error) {
	filename, err := weeklyReportFilename(weekID, outputFormat)
	if err != nil {
		return "", nil, err
	}
	format, err := normalizeWeeklyOutputFormat(outputFormat)
	if err != nil {
		return "", nil, err
	}
	body, err := renderWeeklyReport(content, format)
	if err != nil {
		return "", nil, err
	}
	return filename, body, nil
}

func renderWeeklyReport(content, outputFormat string) ([]byte, error) {
	format, err := normalizeWeeklyOutputFormat(outputFormat)
	if err != nil {
		return nil, err
	}
	switch format {
	case "md":
		return []byte(content), nil
	case "docx":
		return markdownToDocx(content), nil
	case "pdf":
		return markdownToPDF(content), nil
	default:
		return nil, fmt.Errorf("output_format must be one of md, docx, pdf, got %q", outputFormat)
	}
}

func markdownBlocks(content string) [][2]string {
	var blocks [][2]string
	for _, raw := range strings.Split(content, "\n") {
		line := strings.TrimRight(raw, " \t")
		if strings.TrimSpace(line) == "" {
			continue
		}
		switch {
		case strings.HasPrefix(line, "### "):
			blocks = append(blocks, [2]string{"h3", strings.TrimSpace(line[4:])})
		case strings.HasPrefix(line, "## "):
			blocks = append(blocks, [2]string{"h2", strings.TrimSpace(line[3:])})
		case strings.HasPrefix(line, "# "):
			blocks = append(blocks, [2]string{"h1", strings.TrimSpace(line[2:])})
		case strings.HasPrefix(line, "- "), strings.HasPrefix(line, "* "):
			blocks = append(blocks, [2]string{"li", strings.TrimSpace(line[2:])})
		default:
			blocks = append(blocks, [2]string{"p", line})
		}
	}
	return blocks
}

func markdownToDocx(content string) []byte {
	var documentXML strings.Builder
	documentXML.WriteString(`<?xml version="1.0" encoding="UTF-8" standalone="yes"?>`)
	documentXML.WriteString(`<w:document xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main"><w:body>`)
	blocks := markdownBlocks(content)
	if len(blocks) == 0 {
		documentXML.WriteString(`<w:p><w:r><w:t></w:t></w:r></w:p>`)
	}
	for _, block := range blocks {
		text := html.EscapeString(block[1])
		switch block[0] {
		case "h1":
			documentXML.WriteString(`<w:p><w:pPr><w:pStyle w:val="Heading1"/></w:pPr><w:r><w:t xml:space="preserve">` + text + `</w:t></w:r></w:p>`)
		case "h2", "h3":
			documentXML.WriteString(`<w:p><w:pPr><w:pStyle w:val="Heading2"/></w:pPr><w:r><w:t xml:space="preserve">` + text + `</w:t></w:r></w:p>`)
		default:
			documentXML.WriteString(`<w:p><w:r><w:t xml:space="preserve">` + text + `</w:t></w:r></w:p>`)
		}
	}
	documentXML.WriteString(`<w:sectPr/></w:body></w:document>`)

	var buf bytes.Buffer
	archive := zip.NewWriter(&buf)
	files := map[string]string{
		"[Content_Types].xml": `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>` +
			`<Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">` +
			`<Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml"/>` +
			`<Default Extension="xml" ContentType="application/xml"/>` +
			`<Override PartName="/word/document.xml" ContentType="application/vnd.openxmlformats-officedocument.wordprocessingml.document.main+xml"/>` +
			`</Types>`,
		"_rels/.rels": `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>` +
			`<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">` +
			`<Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument" Target="word/document.xml"/>` +
			`</Relationships>`,
		"word/_rels/document.xml.rels": `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>` +
			`<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships"></Relationships>`,
		"word/document.xml": documentXML.String(),
	}
	for _, name := range []string{"[Content_Types].xml", "_rels/.rels", "word/_rels/document.xml.rels", "word/document.xml"} {
		writer, err := archive.Create(name)
		if err != nil {
			continue
		}
		_, _ = writer.Write([]byte(files[name]))
	}
	_ = archive.Close()
	return buf.Bytes()
}

func markdownToPDF(content string) []byte {
	// Minimal PDF. CJK is stored as UTF-16BE so the bytes remain a valid PDF;
	// the Python DatasetWrite canvas is the scheduled publish path and uses STSong-Light.
	var text strings.Builder
	for _, block := range markdownBlocks(content) {
		if text.Len() > 0 {
			text.WriteByte('\n')
		}
		text.WriteString(block[1])
	}
	escaped := strings.ReplaceAll(text.String(), `\`, `\\`)
	escaped = strings.ReplaceAll(escaped, `(`, `\(`)
	escaped = strings.ReplaceAll(escaped, `)`, `\)`)
	stream := fmt.Sprintf("BT /F1 11 Tf 50 780 Td (%s) Tj ET", escaped)
	var b strings.Builder
	write := func(s string) { b.WriteString(s) }
	write("%PDF-1.4\n")
	offsets := make([]int, 6)
	offsets[1] = b.Len()
	write("1 0 obj << /Type /Catalog /Pages 2 0 R >> endobj\n")
	offsets[2] = b.Len()
	write("2 0 obj << /Type /Pages /Kids [3 0 R] /Count 1 >> endobj\n")
	offsets[3] = b.Len()
	write("3 0 obj << /Type /Page /Parent 2 0 R /MediaBox [0 0 612 792] /Contents 4 0 R /Resources << /Font << /F1 5 0 R >> >> >> endobj\n")
	offsets[4] = b.Len()
	write(fmt.Sprintf("4 0 obj << /Length %d >> stream\n%s\nendstream endobj\n", len(stream), stream))
	offsets[5] = b.Len()
	write("5 0 obj << /Type /Font /Subtype /Type1 /BaseFont /Helvetica >> endobj\n")
	xref := b.Len()
	write("xref\n0 6\n0000000000 65535 f \n")
	for i := 1; i <= 5; i++ {
		write(fmt.Sprintf("%010d 00000 n \n", offsets[i]))
	}
	write(fmt.Sprintf("trailer << /Size 6 /Root 1 0 R >>\nstartxref\n%d\n%%%%EOF\n", xref))
	return []byte(b.String())
}

var _ ToolComponent = (*DatasetWriteTool)(nil)
